### DISCUSSION TITLE
Drupal GovCon 2021 Followup Thread

### DISCUSSION THREAD

**Discussion by thisisdano** [uswds/uswds#4355] [General]
👋 Hi to everyone that heard James and me present at Drupal GovCon! 

Let's use this post for any follow-up questions or discussion. As we mentioned in our presentation, we have a new USWDS newsletter coming soon, and you can [sign up for it here](https://public.govdelivery.com/accounts/USGSATTS/subscriber/topics?qsp=GSA_TTS). (When you sign up, you'll be taken to a general Technology Transformation Services email signup form. After you enter your email address, select `USWDS` from the list of subscription topics and select `Submit`. I realize this is a little complicated!)

So, to recap:

### We have a new major version coming in Winter 2022: USWDS 3.0.

This new version will do four things:

- Signal the end of IE11 support.
- Update to modern Sass Module Syntax.
- Allow requiring JavaScript for some functionality.
- Publish semantically versioned component packages.

Here’s what the next release will not do: 

- Move straight to cutting-edge CSS.
- Change markup or look-and-feel.
- Radically change our JavaScript.
- Remove the familiar `uswds` omnibus package.

The USWDS 3.0 beta program will begin in this month: October 2021.

### We want to support informed incremental design.

Starting with USWDS 3.0, we’ll be publishing individual component packages, in addition to the familiar `uswds` omnibus package. Unbundling the design system will allow components to evolve at their own speed and enable projects to tailor their implementation to their unique needs. 

- We want to make it easier to stay up-to-date with the design system.
- We want to prevent major changes to some components from blocking teams from updating to critical changes in other components. 
- We want to make it very clear what’s changed in the design system from version to version so teams can update their implementations intentionally and strategically, at the speed that’s right for their project.
- We want you to know how the design system is changing and how these specific changes will impact your project.
- We’re going to be more strict about semantic versioning to better communicate the impact of the changes we make from version to version.

### We’ll be making it as easy as possible to upgrade and migrate to USWDS 3.0. 

We aim to make an upgrade from the latest version to USWDS 3.0 take minutes, not hours, days, or weeks. Our overall focus is on ease of use over time for designers, developers, and program managers. We learned an important lesson with USWDS 2.0: that migration pain makes any change that much harder.

The biggest migration update, from a developer’s perspective, will probably be related to the Sass Modules work, but we expect that most projects will be able to update only a few lines of code to successfully migrate. Other USWDS conventions (tokens, settings, functions, mixins, and component JavaScript) will continue to work as expected. 

Following along with the USWDS 3.0 beta program will not only help teams better understand what will be changing, but will help the design system make the upgrade process as seamless and painless as possible for everyone. 

### After 3.0: Building real components with a real API 

As we move beyond USWDS 3.0, we'll be experimenting with web components to help bridge the gap between the styles, markup, and guidance we currently provide, and what developers actually need to add USWDS components to their projects. We'll be doing this in public, with support and feedback from the community. 

Providing canonical components could make a big difference in the ease of starting a USWDS project or maintaining a USWDS project over time. There'll certainly be more to come as we move forward!

**Comment by tomsshow** [comment]
How does the flock model influence/affect atomicly designed components?

If I had a `<usa-cta>` that had a `<usa-button>` inside. Will the updates in `<usa-button>` be reflected or referenced in `<usa-cta>`?

**Comment by thisisdano** [comment]
@tomsshow That is the intention, yes.

**Reply by tomsshow** [reply]
Lit vs Stencil, was there a specific reason why stencil is in the running?

Lots of other major design systems seem to have been using lit2

**Reply by mejiaj** [reply]
We're not ruling one or the other out. We're prototyping with both to get a good sense of which one works best for us.

**Reply by tomsshow** [reply]
That sounds perfectly reasonable. We went through a stencil vs lit2 investigation in our org and we found lit2 to be our choice (also it looked like even ionic wasnt using stencil in their internal stuff anymore). Would love to set up a call to talk about it more with your team!

**Comment by btopro** [comment]
if looking for an organization in a very similar space that had to recently go through these decision trees I would recommend connecting with members of the Sakai LMS project. They have a similar design system, with a community of adopters, which uses Sass / utilities and now have to guide that community into a component driven world without upsetting the apple cart along the way. I'd be happy to connect if interested.

Another concept I'd explore is how the assets get there. NARA adopted part of our unbundled methodology a few years back [article](https://dev.to/btopro/part-1-how-penn-state-unbundles-web-components-for-cdn-deployments-20di) for performance and leveraging portions of a portfolio. If distribution of that style is out of scope I'd still check out [Shoelace](https://shoelace.style/) for their CDN based methodology of delivering all approved design assets.

Also if you'd like help prototyping what a component / progressive enhanced version of your library would look like as web components I'm happy to offer feedback. Best of luck.

**Comment by SCourtsMember** [comment]
The Admin Office/U.S. Courts team uses the USWDS on a handful of it's sites and planned to continue using the system on all of our upcoming Drupal 9 websites. Our team is a bit concerned that USWDS will no longer be covering IE 11 in version 3.0. We still have a significant amount of gov users who will be using IE11. I'm curious if there is a process in place to phase out coverage by keeping the current version available for use posted online longer or how other gov sites are handling this change. I'm also open to testing.

**Reply by amstaker** [reply]
Seriously... IE? <<FACEPALM>>

**Comment by maya** [comment]
>Update to modern Sass Module Syntax.

Are you fixing the deprecated Sass issue #4204 and moving to Dart Sass?

**Reply by thisisdano** [reply]
Yes, this will update our codebase to the Sass module system. We already use Dart Sass for compiling (via https://www.npmjs.com/package/sass) but we'll be updating to the modern syntax, and should no longer see the deprecation warnings

**Comment by maya** [comment]
>Publish semantically versioned component packages.

Can you describe the architecture of this (e.g. monorepo, babel)? Or share a link to the branch where this work is being done?

**Reply by thisisdano** [reply]
Still working out the details of this, but will almost certainly be a monorepo

**Comment by mgifford** [comment]
It's not too late to simply resolve this question in a Drupal GovCon 2023 submission https://www.drupalgovcon.org/call-speakers
